### PR TITLE
tools: Move ui-metalk8s-client-generator to tools/crd-client-generator-js and refactor it to be completely uncoupled with metalk8s

### PR DESCRIPTION
**Pull request by JBWatenbergScality**
**Component**:

tools, crd-client-generator

**Context**: 

Move ui-metalk8s-client-generator to tools/crd-client-generator-js.
I also introduced a new optional cli parameter `typePrefix` which was before hardcoded to Metalk8s, coupling the crd generator with metalk8s whereas it could be used for any other crd.

### CODE REVIEW

**Review by gdemonet — DISMISSED** [review]

File: eve/main.yml [hunk 1]
```diff
@@ -1336,6 +1336,49 @@ stages:
             STEP_NAME: unit_tests_ui
       - Upload: *upload_final_status_artifact
 
+  unit_tests_crd_client_generator:
```
Comment (gdemonet) [inline eve/main.yml:1314]: You need to add this stage to the stages triggered by `unit_tests`

File: eve/main.yml [hunk 2]
```diff
@@ -1336,6 +1336,49 @@ stages:
...
+            <<: *_env_final_status_artifact_failed
+            <<: *_unit-test_crd_client_generator_junit_info
+            STEP_NAME: unit_tests_crd_client_generator
+      - ShellCommand: *setup_cache
+      - ShellCommand:
+          name: Install CRD client generator dependencies
+          workdir: build/tools/crd-client-generator-js
+          command: >
+            npm ci --prefer-offline --no-audit
+          haltOnFailure: true
```
Comment (gdemonet) [inline eve/main.yml:1341]: Could be done in the worker image, to cache those dependencies as much as possible

File: eve/main.yml [hunk 3]
```diff
@@ -1336,6 +1336,49 @@ stages:
...
+    _metalk8s_internal_info:
+      junit_info: &_unit-test_crd_client_generator_junit_info
+        TEST_SUITE: unit test
+        TEST_NAME: crd client generator
+    worker:
+      type: kube_pod
+      path: eve/workers/pod-unit-tests/pod.yaml
+      images:
+        docker-unit-tests:
+          context: 'ui'
```
Comment (gdemonet) [inline eve/main.yml]: not sure it makes sense to use this build context

File: eve/main.yml (hunk 2 above)
Comment (JBWatenbergScality) [reply eve/main.yml:1341]: This is already done in the worker : https://github.com/scality/metalk8s/blob/development/2.8/eve/workers/pod-unit-tests/ui/Dockerfile#L35 but the build context was targeting `ui` instead of `tools/crd-client-generator-js` so wrong dependencies were cached.

File: tools/crd-client-generator-js/package.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
     "@types/jest": "^26.0.19",
     "babel-jest": "^26.6.3",
     "jest": "^26.6.3",
-    "ts-node": "^9.1.1"
+    "ts-node": "^9.1.1",
+    "typescript": "^4.1.3"
```
Comment (gdemonet) [inline tools/crd-client-generator-js/package.json:26]: How is this new dependency related to the "move" commit?

File: tools/crd-client-generator-js/README.md [hunk 5]
```diff
@@ -1,11 +1,17 @@
-## Metalk8s client generator
+## CRD client generator
 
-It generates a flow client based on `@kubernetes/client` custom objects client from kubernetes `CustomResourceDefinition` of metalk8s volumes.
+It generates a flow client based on `@kubernetes/client` custom objects client from a kubernetes `CustomResourceDefinition`.
```
Comment (gdemonet) [inline tools/crd-client-generator-js/README.md]: minor: capitalize "Kubernetes"

File: eve/main.yml (hunk 2 above)
Comment (gdemonet) [reply eve/main.yml:1341]: Is this still required then?

File: tools/crd-client-generator-js/src/generateClient.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 const YAML = require('yaml');
 const fs = require('fs');
+import { version } from '@babel/core';
```
Comment (gdemonet) [inline tools/crd-client-generator-js/src/generateClient.ts]: Automatic import? Don't think this one is used :/

File: tools/crd-client-generator-js/src/generateClient.ts [hunk 7]
```diff
@@ -17,15 +18,29 @@ export function generateClient(crdFile: string, destinationFile: string, prefix?
...
-  const typePrefix = (prefix || '') + crdSpec.spec.version.replace(/^\w/, (c) => c.toUpperCase());
+  const generateClientHeader = `
+  //@flow
+  // /!\\ WARNING :  THIS FILE IS AUTOGENERATED FROM A KUBERNETES CUSTOM RESOURCE DEFINITION FILE. DO NOT CHANGE IT, use crd-client-generator-js to update it.
+  import { customObjects } from './api';
+  export type Result<T> = T | {error: any};
+  
+  `;
+
+  const versionsClients = versions.map(version => {
```
Comment (gdemonet) [inline tools/crd-client-generator-js/src/generateClient.ts]: New block, but it looks like the code that follows was not indented

**Review by gdemonet — CHANGES_REQUESTED** [review]
Some (late, sorry) comments

File: tools/crd-client-generator-js/package.json (hunk 4 above)
Comment (JBWatenbergScality) [reply tools/crd-client-generator-js/package.json:26]: This is not related to the move I found it was missing when reinstalling dependencies, I will move this to a separate commit

File: tools/crd-client-generator-js/src/generateClient.ts [hunk 8]
```diff
@@ -1,59 +1,90 @@
...
+  const file = fs.readFileSync(crdFile, "utf8");
   const crdSpec = YAML.parse(file);
-  
+
   const generator = new Generator();
+
+  const versions =
+    crdSpec.apiVersion === "apiextensions.k8s.io/v1"
+      ? crdSpec.spec.versions.map((version) => version.name)
+      : [crdSpec.spec.version];
```
Comment (gdemonet) [inline tools/crd-client-generator-js/src/generateClient.ts]: Sorry, just caught this: I don't think the `versions` field only exists in `v1`. It's the opposite: in `v1beta1`, there is **also** a `version` field, and both may be specified (in which case `version` **must** match `versions[0]`).

File: tools/crd-client-generator-js/src/generateClient.ts [hunk 9]
```diff
@@ -1,59 +1,90 @@
...
+  const jsonSchemas =
+    crdSpec.apiVersion === "apiextensions.k8s.io/v1"
+      ? crdSpec.spec.versions.reduce(
+          (aggregatedJsonSchemas, version) => ({
+            ...aggregatedJsonSchemas,
+            [version.name]: version.schema.openAPIV3Schema,
+          }),
+          {}
+        )
+      : { [crdSpec.spec.version]: crdSpec.spec.validation.openAPIV3Schema };
```
Comment (gdemonet) [inline tools/crd-client-generator-js/src/generateClient.ts]: Same here: `spec.validation` may describe a top-level schema for all versions, but it's an optional field, and `v1beta1` CRDs can use `spec.versions[*].schema` instead.

**Review by gdemonet — CHANGES_REQUESTED** [review]
Oops, missed some important details last time (that's why we try not to mix commits which update logic with ones that only apply formatting changes, otherwise reviewers may not be careful enough :wink:).
For reference:
- [CRD v1beta1](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#customresourcedefinition-v1beta1-apiextensions-k8s-io)
- [CRD v1](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#customresourcedefinition-v1-apiextensions-k8s-io)

Maybe you want to use `@kubernetes/client` types in your generator? :stuck_out_tongue_winking_eye:

**Comment by JBWatenbergScality** [comment]
I should have thought about using `@kubernetes/client` types here before ! Thanks 👍   !

File: tools/crd-client-generator-js/src/generateClient.ts [hunk 10]
```diff
@@ -1,59 +1,95 @@
...
+    ? { [crdSpec.spec.version]: crdSpec.spec.validation.openAPIV3Schema }
+    : (crdSpec.spec.versions || []).reduce(
+        (aggregatedJsonSchemas, version) => ({
+          ...aggregatedJsonSchemas,
+          [version.name]:
+            version.schema?.openAPIV3Schema ||
+            crdSpec.spec.validation.openAPIV3Schema,
+        }),
+        {}
+      );
```
Comment (gdemonet) [inline tools/crd-client-generator-js/src/generateClient.ts:43]: :+1:

**Review by gdemonet — DISMISSED** [review]

File: eve/main.yml (hunk 2 above)
Comment (gdemonet) [reply eve/main.yml:1341]: Seen offline: it's required since we can't cache `node_modules` in the image (as we need to mount the sources over it). Though the worker build should pre-install them, saving as much as possible in `~/.npm`, and this step tries to re-use this cache as much as possible with `--prefer-offline`.

**Review by ChengYanJin — APPROVED** [review]
